src/arithmetic/fields.rs

```diff
@@ -103,6 +140,197 @@ pub trait FieldExt:
+    ///
+    /// where ROOT_OF_UNITY is a generator of the order 2^n subgroup (and therefore a nonsquare).
+    ///
+    /// The choice of root from sqrt is unspecified.
+    pub fn sqrt_ratio(&self, num: &F, div: &F) -> (Choice, F) {
+        // Based on:
+        // * [Sarkar2020](https://eprint.iacr.org/2020/1407)
+        // * [BDLSY2012](https://cr.yp.to/papers.html#ed25519)
+        //
+        // We need to calculate uv and v, where v = u^((m-1)/2), u = num/div, and p-1 = T * 2^S.
+        // We can rewrite as follows:
+        //
+        //      v = (num/div)^((T-1)/2)
+        //        = num^((T-1)/2) * div^(p-1 - (T-1)/2)    [Fermat's Little Theorem]
+        //        =       "       * div^(T * 2^S - (T-1)/2)
+        //        =       "       * div^((2^(S+1) - 1)*(T-1)/2 + 2^S)
+        //        = (num * div^(2^(S+1) - 1))^((T-1)/2) * div^(2^S)
+        //
+        // Let  w = (num * div^(2^(S+1) - 1))^((T-1)/2) * div^(2^S - 1).
+        // Then v = w * div, and uv = num * v / div = num * w.
+        //
+        // We calculate:
+        //
+        //      s = div^(2^S - 1) using an addition chain
+        //      t = div^(2^(S+1) - 1) = s^2 * div
+        //      w = (num * t)^((T-1)/2) * s using another addition chain
+        //
+        // then u and uv as above. The addition chains are given in
+        // https://github.com/zcash/pasta/blob/master/addchain_sqrt.py .
+        // The overall cost of this part is similar to a single full-width exponentiation,
+        // regardless of S.
+
+        let sqr = |x: F, i: u32| (0..i).fold(x, |x, _| x.square());
+
+        // s = div^(2^S - 1)
+        let s = (0..5).fold(*div, |d: F, i| sqr(d, 1 << i) * d);
+
+        // t == div^(2^(S+1) - 1)
+        let t = s.square() * div;
+
+        // w = (num * t)^((T-1)/2) * s
+        let w = (t * num).pow_by_t_minus1_over2() * s;
+
+        // v == u^((T-1)/2)
+        let v = w * div;
+
+        // uv = u * v
+        let uv = w * num;
+
+        let res = self.sqrt_common(&uv, &v);
+
+        let sqdiv = res.square() * div;
+        let is_square = (sqdiv - num).ct_is_zero();
+        let is_nonsquare = (sqdiv - F::ROOT_OF_UNITY * num).ct_is_zero();
+        assert!(bool::from(
+            num.ct_is_zero() | div.ct_is_zero() | (is_square ^ is_nonsquare)
+        ));
+
+        (is_square, res)
+    }
+
+    /// Same as sqrt_ratio(u, one()) but more efficient.
+    pub fn sqrt_alt(&self, u: &F) -> (Choice, F) {
+        let v = u.pow_by_t_minus1_over2();
+        let uv = *u * v;
+
+        let res = self.sqrt_common(&uv, &v);
+
+        let sq = res.square();
+        let is_square = (sq - u).ct_is_zero();
+        let is_nonsquare = (sq - F::ROOT_OF_UNITY * u).ct_is_zero();
+        assert!(bool::from(u.ct_is_zero() | (is_square ^ is_nonsquare)));
+
+        (is_square, res)
+    }
+
+    /// Common part of sqrt_ratio and sqrt_alt: return res given v = u^((T-1)/2) and uv = u * v.
+    fn sqrt_common(&self, uv: &F, v: &F) -> F {
```

Comment:
I have not checked this function yet.

Reply:
I checked that this function matches `SqrtField.sarkar_sqrt_common()` in [`squareroottab.sage`](https://github.com/zcash/pasta/blob/master/squareroottab.sage).